suitesparse_bindings/sprs_suitesparse_umfpack/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+            /// with a stored C handle to an efficient (but opaque) direct solver.
+            ///
+            /// This factorization can be done for either square or rectangular `A` matrix,
+            /// but can only be solved directly if `A` is square.
+            pub fn new<N>(a: CsMatI<N, $int>) -> Self
+            where N: Default + Clone + Into<f64>,
+            {
+                // UMFPACK methods require CSC format and f64 data
+                let a: CsMatI<f64, $int> = a.to_other_types().into_csc();
+
+                // Get shape info
+                let nrow_ = a.rows();
+                let ncol_ = a.cols();
+                let nnz = a.nnz();
+
+                // Get C-compatible raw pointers to column pointers, indices, and values
+                let ap = a.indptr().to_proper().as_ptr();
+                let ai = a.indices().as_ptr();
+                let ax = a.data().as_ptr();
+
+                // Do symbolic factorization
+                let symbolic_inner = &mut (null_mut() as *mut c_void) as *mut *mut c_void;
+                unsafe {
+                    $symbolic(
+                        nrow_ as $int,
+                        ncol_ as $int,
+                        ap,
+                        ai,
+                        ax,
+                        symbolic_inner,
+                        null() as *const f64,  // Default settings
+                        null_mut() as *mut c_void  // Ignore info
+                    );
+                };
+                let symbolic_ = unsafe{$Symbolic(*symbolic_inner)};
+
+                // Do numeric factorization
+                let numeric_inner = &mut (null_mut() as *mut c_void) as *mut *mut c_void;
+                unsafe {
+                    $numeric(
+                        ap,
+                        ai,
+                        ax,
+                        symbolic_.0,
+                        numeric_inner,
+                        null() as *const f64,  // Default settings
+                        null_mut() as *mut c_void  // Ignore info
+                    );
+                };
+                let numeric_ = unsafe{$Numeric(*numeric_inner)};
+
+                Self {
+                    _a: a,
+                    symbolic: symbolic_,
+                    numeric: numeric_,
+                    nrow: nrow_,
+                    ncol: ncol_,
+                    _nnz: nnz
+                }
+            }
+
+            /// Get the shape of A like (`nrow`, `ncol`)
+            pub fn shape(&self) -> (usize, usize) {
+                (self.nrow as usize, self.ncol as usize)
+            }
+
+            /// Get the number of nonzero entries in A
+            pub fn nnz(&self) -> usize {
+                self._nnz as usize
+            }
+
+            /// Get a reference to the stored matrix,
+            /// which may have had its data type converted from
+            /// what was supplied.
+            pub fn a(&self) -> &CsMatI<f64, $int> {
```

Comment:
It is more ergonomic to return `CsMatViewI<f64, $int>` (use `self.a.view()` in the function body)

Reply:
Done